### ISSUE TITLE
Arbitrary header levels like h7, h8, h9, etcl that show on the ToC

### ISSUE THREAD

**Issue by cirosantilli** [asciidoctor/asciidoctor#3358]
I know, core is too big, create an extension, why would anyone ever need that, yada yada XD (I'm kidding Dan, I love you).

Let's summarize here investigations on the best way of achieving it with an extension or some other feature.

Could just render as `<div class="header-7"`.

Related: http://discuss.asciidoctor.org/Heading-7-not-handled-td710.html

```
= Heading 1
:idprefix:
:idseparator: -
:sectanchors:
:sectlinks:
:sectnums:
:toc: preamble
:toclevels: 999

== Heading 2

=== Heading 3

==== Heading 4

===== Heading 5

====== Heading 6

======= Heading 7

======== Heading 8

========= Heading 9

Deeeep.
```

**Comment by mojavelinux** [comment]
When something similar came up with lists, I first resisted, only to realize it was no extra processing to add arbitrarily deep levels. So we did it.

However, headings are somewhat different. The reason is, HTML doesn't have arbitrary heading depths. So the question becomes, what do we map it too? Of course, we could just cut the HTML off at h6 and just keep repeating that tag indefinitely. Since it gets wrapped in the section level div, you don't necessary lose the information.

However, I really strongly argue that if you are going beyond 6 levels of sections, you probably need to rethink the structure of your document. Maybe a computer can read it, but not a human.

Regardless, I still think in this case, this is AsciiDoc spec material. There's a reason HTML stops at h6 and it isn't an arbitrary choice.

**Comment by mojavelinux** [comment]
So I pose you this question. What's the use case?

**Comment by elextr** [comment]
Well I have seen a (not made with Asciidoc) document with 8 levels, but it was a 2000 page formal specification.  So there is a use-case, but its rare.

**Comment by mojavelinux** [comment]
So it seems the argument is more about portability with legacy / migrated documents than it is about actually needing the structure for writing. I can certainly see that.

Was that 2000 page document rendered to HTML? If so, what happened with those section titles? (Or was it just a PDF document?)

**Comment by elextr** [comment]
I only saw PDF.

**Comment by cirosantilli** [comment]
Output: `<div class="header-7"`, I had edited it in description before you replied :-)

Use case: I already have to split my documents artificially to avoid the h6 barrier. E.g.:

- https://github.com/cirosantilli/linux-kernel-module-cheat/tree/9763f3ccf007983bbc432adbc7c157f0dff106bb#userland-content
- https://github.com/cirosantilli/linux-kernel-module-cheat/tree/9763f3ccf007983bbc432adbc7c157f0dff106bb#userland-assembly
- https://github.com/cirosantilli/linux-kernel-module-cheat/tree/9763f3ccf007983bbc432adbc7c157f0dff106bb#arm-userland-assembly

ARM userland assembly fits in perfectly in Userland content.

But I had to split it twice already because I have an h6 in there: https://github.com/cirosantilli/linux-kernel-module-cheat/tree/9763f3ccf007983bbc432adbc7c157f0dff106bb#arm-loop-over-array

No depth limit is deep enough.

**Comment by mojavelinux** [comment]
> No depth limit is deep enough.

This is categorically untrue. That's why we have the field of information architecture and writing style guidelines like ALA and Chicago manual of style. If you were writing for a computer to read it, depth might not matter. But in information system geared for humans, there are decades of research as to why limits matter. So you're throwing out a whole field of information science by making that claim.

I get that there extreme circumstances that put us over the limit of what's possible right now. And we can look at those. But a limit is essential.

Generally speaking, if content is blowing through these limits (which, as I said, might be overly restrictive atm), then it's the content that is problematic.

**Comment by mojavelinux** [comment]
You can't justify a change to the language just because the structure of the content is bad. That's like saying that you want to change a programming language because the code is inefficient or buggy. It's just not a justifiable argument.

**Comment by mojavelinux** [comment]
(I'm not implying your content is bad. But if someone said they had 10+ section levels, then that would be...or it's abusing the structure for a purpose it wasn't designed).

**Comment by elextr** [comment]
> then it's the content that is problematic.

The use-case tends to be formal documents where referencability is more important than readability.  In those cases style is sacrificed to allow a more important capability.  Its unusual content, but the only place its "problematic" is for the tools.

Given that the situation is unusual I would think @mojavelinux would be correct if he said its an extension to generate output for that, especially in HTML which has a specific structural limit.  So long as its not limited by the frontend, then backends or extensions may apply whatever limit is appropriate, and it seems perfectly reasonable to me for the standard HTML backend to stick with 6.

**Comment by mojavelinux** [comment]
In this case, it requires more than an extension since it's really a parser concern. But what might be reasonable is to allow the number of levels captured to be configured (perhaps through an AsciiDoc attribute or compliance flag). Then, it *would* be up to you to handle them in the output using a custom converter. Though, as it stands right now, both the HTML and DocBook converters will happy just keep recursing. In HTML, you'll end up with h7, h8, h9, etc.

So I think we could explore where this maximum level is controlled. I like a compliance setting (though that makes it unavailable on GitHub...but the result would be stripped away anyway since GitHub also enforces the limit).

**Comment by elextr** [comment]
> Though, as it stands right now, both the HTML and DocBook converters will happy just keep recursing. In HTML, you'll end up with h7, h8, h9, etc.

Yes, if the frontend doesn't limit it, then the backend needs to.  But it does allow backends which have limits to apply the ones they need, otherwise it needs the frontend to limit everybody to the lowest common denominator.

Whats a "compliance setting"?  

As its a property of the document that it uses a specific number of levels, it would be best if the setting was controlled by the document so it doesn't get forgotten.

**Comment by cirosantilli** [comment]
I'm fine with `<h6>` all the way down too (although I think that is a bit confusing), or anything else that shows on ToC and can be styled with CSS.

The structure of my content is pristine, and no one can convince me otherwise!!! :laughing: :laughing: :laughing: :laughing: :laughing: :laughing: :laughing: :laughing:

**Comment by robszewczyk** [comment]
Hello!  My first time participating in Asciidoctor development.  We are using the Asciidoctor in one of the usecases outlined above (sizeable formal specification) and we've run into the section depth limitations.  I've attached a patch that addresses some of the concerns raised: I did not go to the next level and made the section depth unlimited, but I restricted it to 9.  If such a patch is desirable in AsciiDoc, I'd be more than happy to make appropriate changes.

**Comment by mojavelinux** [comment]
The plan at the moment is to declare in the AsciiDoc Language specification that sections can be nested to an arbitrary depth. It will then be the responsibility of converters to decide how to handle that situation.

Since this is a language change, we can't make the change in Asciidoctor until it the proposal is accepted by the AsciiDoc Language project (which means it has an SDR, see https://gitlab.eclipse.org/eclipse/asciidoc-lang/asciidoc-lang/-/tree/main/spec/sdrs).

In the new HTML converter, the plan is to use `<section>` tags for nested sections (much like DocBook). Once arbitrary section levels are permitted, it will then need to represent the headings that are deeper than h6 using an alternate form, such as `<p class="h7">...<p>` or `<h6 class="h7">...</h6>`. The fact still remains that HTML has no way of representing h7 and beyond natively.

**Comment by mojavelinux** [comment]
For what it's worth, the DocBook toolchain just sticks with `<h6>` tags for every nested section beyond the first level that uses the h6 tag. So that's a precedent from which we can build. The DocBook toolchain doesn't add a class to differentiate the h6 tags, though I think that would be nice to have.
